File: services/python-services/observability/business_health.py
from __future__ import annotations

import os
from datetime import UTC, datetime

import httpx

from observability.schemas import BusinessHealthMetrics


async def get_business_health_metrics() -> BusinessHealthMetrics:
    """
    Proxies to the Node.js backend /api/internal/metrics/business-health.
    Falls back to source='unavailable' — never invents production values.
    """
    backend_url = os.getenv("BACKEND_URL", "").rstrip("/")
    api_key = os.getenv("MONITORING_API_KEY", "")

    if not backend_url:
        return BusinessHealthMetrics(
            last_checked_at=datetime.now(UTC),
            source="unavailable",
        )

    headers: dict[str, str] = {"Content-Type": "application/json"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"

    try:
        async with httpx.AsyncClient(timeout=8) as client:
            resp = await client.get(
                f"{backend_url}/api/internal/metrics/business-health",
                headers=headers,
            )

        if not resp.is_success:
            return BusinessHealthMetrics(
                last_checked_at=datetime.now(UTC),
                source="unavailable",
            )

        data = resp.json()
        return BusinessHealthMetrics(
            active_tenants=data.get("activeTenants"),
            suspended_tenants=data.get("suspendedTenants"),
            trialing_tenants=data.get("trialingTenants"),
            shipments_created_today=data.get("shipmentsCreatedToday"),
            shipments_in_transit=data.get("shipmentsInTransit"),
            stuck_shipments=data.get("stuckShipments"),
            last_checked_at=datetime.now(UTC),
            source="real",
        )
    except Exception:
        return BusinessHealthMetrics(
            last_checked_at=datetime.now(UTC),
            source="unavailable",
        )

File: apps/status-dashboard/services/business-health.js
const BACKEND_URL    = process.env.BACKEND_URL    ?? 'http://localhost:3001';
const MONITORING_KEY = process.env.MONITORING_API_KEY ?? '';

let businessHealth = {
  activeTenants:        null,
  suspendedTenants:     null,
  trialingTenants:      null,
  shipmentsCreatedToday: null,
  shipmentsInTransit:   null,
  stuckShipments:       null,
  isMock:               false,
  available:            false,
  lastChecked:          null,
  error:                null,
};

export function getBusinessHealth() {
  return businessHealth;
}

export async function refreshBusinessHealth() {
  try {
    const headers = { 'Content-Type': 'application/json' };
    if (MONITORING_KEY) headers['Authorization'] = `Bearer ${MONITORING_KEY}`;

    const res = await fetch(`${BACKEND_URL}/api/internal/metrics/business-health`, {
      headers,
      signal: AbortSignal.timeout(8000),
    });

    if (!res.ok) {
      businessHealth = { ...businessHealth, available: false, error: `HTTP ${res.status}`, lastChecked: new Date().toISOString() };
      return;
    }

    const data = await res.json();
    businessHealth = {
      ...data,
      available:   true,
      error:       null,
      lastChecked: new Date().toISOString(),
    };
  } catch (err) {
    businessHealth = {
      ...businessHealth,
      available:   false,
      error:       err.message,
      lastChecked: new Date().toISOString(),
    };
  }
}
